lib/src/license_detection/confidence.dart

```diff
@@ -0,0 +1,236 @@
+      diffs,
+      Range(range.start, range.end),
+    );
+  }
+
+  // If confidence is not greater than or equal to threshold don't
+  // return any match.
+  return null;
+}
+
+/// Calculates the confidence percentange as 1 subtracted by the ratio of
+/// Levenshtein word [distance] to the number of tokens in known license.
+@visibleForTesting
+double confidencePercentage(int knownLength, int distance) {
+  if (knownLength == 0) {
+    return 1.0;
+  }
+
+  return 1.0 - (distance / knownLength);
+}
+
+@visibleForTesting
+List<Diff> getDiffs(
+  List<Token> inputTokens,
+  List<Token> knownTokens,
+  MatchRange matchRange,
+) {
+  final dmp = DiffMatchPatch();
+
+  final unknownText = tokensNormalizedValue(inputTokens
+      .skip(matchRange.input.start)
+      .take(matchRange.input.end - matchRange.input.start));
+
+  final knownText = tokensNormalizedValue(
+    knownTokens.take(matchRange.source.end),
+  );
+
+  final diffs = dmp.diffMain(unknownText, knownText);
+
+  dmp.diffCleanupSemantic(diffs);
+
+  return List.unmodifiable(diffs);
+}
+
+/// Returns the range in [diffs] that best resembles the known license text.
+///
+/// The range provides diffs from which the unknown text could be trimmed down to
+/// produce best resemble known text. Essentialy it tries to remove part's of
+/// text in unknown license which are not a part of known license without affecting the
+/// confidence negatively i.e any false-negatives are not discarded while some of the
+/// true-negatives are discarded.
+@visibleForTesting
+Range diffRange(String known, List<Diff> diffs) {
+  var foundStart = false;
+  var seen = '';
+  var end = 0;
+  var start = 0;
+
+  for (end = 0; end < diffs.length; end++) {
+    if (seen.length > 1 && seen.trimRight() == known) {
+      break;
+    }
+
+    if (diffs[end].operation == Operation.insert ||
+        diffs[end].operation == Operation.equal) {
+      if (!foundStart) {
+        start = end;
+        foundStart = true;
+      }
+
+      seen += diffs[end].text + ' ';
+    }
+  }
+
+  return Range(start, end);
+}
+
+/// Returns a score indicating to what extent the [diffs] are acceptable.
+///
+/// If a negative integer is returned it implies that the changes are not
+/// accepatable and hence we discard the known license against which the diffs
+/// were calculated.
+///
+/// If a positive integer is returned it indicates the Levenshtein edit
+/// distance calculated on the number of words altered instead of
+/// the number of characters altered.
+@visibleForTesting
+int scoreDiffs(Iterable<Diff> diffs) {
+  var prevText = '';
+  var prevDelete = '';
+
+  /// Check if there was a version change and return true accordingly.
+  bool verifyVersionChange(String number) {
+    if (_isVersionNumber(number) && prevText.endsWith('version')) {
+      if (!prevText.endsWith('the standard version') &&
+          !prevText.endsWith('the contributor version')) {
+        return true;
+      }
+    }
+
+    return false;
+  }
```

Comment:
Can we not make this a top-level function that takes `Iterable<Diff> diffs` as an argument? and throws an exception?

Reply:
```
verifyNoVersionChange(Iterable<Diff> diffs) {
  var prevText = '';
  var prevDelete = '';

    for (var diff in diffs) {
    final text = diff.text;

....
}
```